In [1]:
import pandas as pd
import csv

- The first code line is going to convert the csv document in wich i've stored all the name of the song and album
mentionned in a subbreddit about HipHop
- to clean the data i've converted the csv file to a dataframe and since my column in the song.csv were not named,
I did it to have a better view of the data meaning
- df.describe() and df.info() is a function used to give a global view of the datas
- df.head() is a function used to view a certain number of row from the data table in this case I chosed to view
the 15 first row of the Data table to be sure that my dataframe was correctly imported

In [2]:
df = pd.read_csv("song.csv", header=None)
df.columns = ["Type", "Song", "created_utc", "score", "url"]
df.describe()
df.info()
df.head(15)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 606 entries, 0 to 605
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Type         606 non-null    object 
 1   Song         329 non-null    object 
 2   created_utc  606 non-null    float64
 3   score        606 non-null    int64  
 4   url          606 non-null    object 
dtypes: float64(1), int64(1), object(3)
memory usage: 23.8+ KB


Type                                               Song   created_utc  \
0   ALBUM  #Albums\n\n- [**The Weeknd - Dawn FM**](https:...  1.641577e+09   
1   ALBUM  Album of the Year 2021 #24: DMX - Exodus\n\nAr...  1.641579e+09   
2   ALBUM  Earl Sweatshirt shared the tracklist for his n...  1.641584e+09   
3   ALBUM                                                NaN  1.641586e+09   
4   ALBUM  [APPLE MUSIC](https://music.apple.com/us/album...  1.641531e+09   
5    SONG  [FRESH VIDEO] The Weeknd - Sacrifice (Official...  1.641586e+09   
6   ALBUM                                                NaN  1.641592e+09   
7   ALBUM                                                NaN  1.641532e+09   
8   ALBUM  Happy 2022 /r/hiphopheads! Hope you all enjoye...  1.641575e+09   
9    SONG                   [FRESH] Earl Sweatshirt- Titanic  1.641532e+09   
10  ALBUM  I would start with my favourite which is Saigo...  1.641553e+09   
11  ALBUM                                                NaN  1.641603e+09   
12   SONG  [FRESH] Lupe Fiasco spits over Streets of Rage...  1.641551e+09   
13  ALBUM  **Welcome to the /r/hiphopheads daily discussi...  1.641583e+09   
14   SONG          [FRESH VIDEO] SZA - I Hate U (Visualizer)  1.641577e+09   

    score                                                url  
0      94  https://www.reddit.com/r/hiphopheads/comments/...  
1      63  https://www.reddit.com/r/hiphopheads/comments/...  
2    1485  https://www.reddit.com/r/hiphopheads/comments/...  
3     312                       https://youtu.be/UhD-Bs7k64c  
4    5392  https://www.reddit.com/r/hiphopheads/comments/...  
5     265                       https://youtu.be/VafTMsrnSTU  
6     115        https://www.youtube.com/watch?v=p8rBpvADP5o  
7    2158  https://open.spotify.com/album/4oxmme6i4mypSt2...  
8     216  https://www.reddit.com/r/hiphopheads/comments/...  
9    1753  https://music.apple.com/ca/album/titanic/15991...  
10    387  https://www.reddit.com/r/hiphopheads/comments/...  
11     26  https://www.nme.com/news/music/dj-kay-slay-has...  
12    394                    https://v.redd.it/yvg52853r8a81  
13     61  https://www.reddit.com/r/hiphopheads/comments/...  
14     68        https://www.youtube.com/watch?v=xEAEDKzjQTY

- df.info() informed us that only 329 of the 606 entries in the song categories are non-null
- In the following code i'm going to create a new Data Frame with only non-null entries


In [3]:
df1 = df.dropna()
print(df1)
df1.head()
df1.info()

      Type                                               Song   created_utc  \
0    ALBUM  #Albums\n\n- [**The Weeknd - Dawn FM**](https:...  1.641577e+09   
1    ALBUM  Album of the Year 2021 #24: DMX - Exodus\n\nAr...  1.641579e+09   
2    ALBUM  Earl Sweatshirt shared the tracklist for his n...  1.641584e+09   
4    ALBUM  [APPLE MUSIC](https://music.apple.com/us/album...  1.641531e+09   
5     SONG  [FRESH VIDEO] The Weeknd - Sacrifice (Official...  1.641586e+09   
..     ...                                                ...           ...   
598   SONG  [ORIGINAL] ethan h - Swirling and Falling (Add...  1.640911e+09   
601  ALBUM  Nas’ eighth studio album continues the run of ...  1.640805e+09   
602   SONG  [FRESH] C Blu - A Christmas Carol (feat. 26AR)...  1.640855e+09   
604  ALBUM  The [trailer](https://youtu.be/5QLFrE2qSZo) is...  1.626709e+09   
605  ALBUM  My Turn is the second studio album by American...  1.614533e+09   

     score                                         

- Since All of my Data is non-null i can start working on it
- I have 2 type of Data wich is Song and Album
- The song type is easier to work with because it only got one value
- Now i'm going to see how many rows type are song (251 rows)

In [4]:
df_song = df1[df1["Type"]=="SONG"]

-Now i'm going to see how many rows type are album (75 rows)

In [5]:
df_album = df1[df1["Type"]=="ALBUM"]

- Now I'm going to put all the song name of type song mentionned in the subreddit in a list

In [6]:
song_list = df_song['Song'].to_list()
print(song_list)
print(type(song_list))

['[FRESH VIDEO] The Weeknd - Sacrifice (Official Music Video)', '[FRESH] Earl Sweatshirt- Titanic', '[FRESH] Lupe Fiasco spits over Streets of Rage 2 Player Select Music', '[FRESH VIDEO] SZA - I Hate U (Visualizer)', '[FRESH] The Weeknd - Less Than Zero', '[FRESH VIDEO] ZelooperZ - Each and Every Moment (Prod. Dilip )', '[FRESH VIDEO] Gunna - too easy Remix (feat. Future & Roddy Ricch)', '[FRESH] 2 Chainz - Million Dollars Worth of Game (feat. 42 Dugg)', '[FRESH] The Weeknd - Is There Someone Else?', '[FRESH] The Weeknd ft. Lil Wayne - I Heard You’re Married (Official Audio)', '[FRESH] iLoveMakonnen & Yellow Trash Can - My Girl Trans', '[FRESH VIDEO] Big Cheeko (feat. Mach-Hommy) - Spin Off (Official Music Video)', '[FRESH VIDEO] Slug Christ- Horribly Wrong', '[FRESH] FROGGY FRESH - The Last Shot', '[FRESH] Da$h - 10,000 Hours', '[FRESH VIDEO] SAINt JHN - The Best Part Of Life (Live on Colbert)', '[FRESH] Azizi Gibson - Hate To Say It (feat. Freddie Gibbs)', '[FRESH] Kai Bandz - Gettin

- Now I'm going to put all the song name of type Album mentionned in the subreddit in a list

In [7]:
album_list = df_album['Song'].to_list()
print(album_list)
print(type(album_list))



["#Albums\n\n- [**The Weeknd - Dawn FM**](https://music.apple.com/us/album/dawn-fm/1603171516) (featuring Lil Wayne, Tyler the Creator, Quincy Jones, Oneohtrix Point Never & Jim Carrey) [R&B, Pop, 80s Pastiche, *Republic / XO*]\n- [**Gunna - DS4EVER**](https://music.apple.com/us/album/ds4ever/1603748074) (featuring Drake, Lil Baby, 21 Savage, Future, Young Thug, Kodak Black, G Herbo, Chris Brown, Yung Bleu, Chlöe Bailey, Nechie & Roddy Ricch) [Trap, ATL, *YSL / 300*]\n- [**FNKPMPN (Del The Funky Homosapien & Kool Keith) - Subatomic**](https://koolkeith.bandcamp.com/album/subatomic) (featuring Lil B & Q-Shy) [Alt-Rap, *Indie*]\n- [**Sentino - Aporofobia**](https://music.apple.com/us/album/aporofobia/1603788908) [Polish Rap, *Sicarios*]\n- [**Montana of 300 - Rap God**](https://music.apple.com/us/album/rap-god/1598273735) [Peoria Trap, *FGE*]\n- [**DJ Muggs & Rigz - Gold**](https://soulassassins.com/)* [Boom Bap, *Soul Assassins*]\n- [**OG Keemo - Mann beisst Hund**](https://open.spotify

- I will convert into a csv document each of the following list
1. the first document is a csv file with the name of all the song who are not of type album
2. the second document is a csv file with the name of all the song who are part of an album
3. This is file is for personal practice and is going to be a combined file of all type of songs


In [8]:
#document 1
single = df_song.to_csv("Single.csv")
album = df_album.to_csv("Album.csv")
musique = df1.to_csv("Api.csv")

I'm going to do the cleaning of the single.csv from the dataframe song_df. (I'm creating single.csv before
single_clean.csv because as an practice purpose I want to see the transformation of my data but I could have gone
directly to single_clean.csv without passing through single.csv)

In [9]:
remove_words = ["FRESH", "FRESH VIDEO", "VIDEO", "Hype", "ORIGINAL", "THROWBACKTHURSDAY", "OfficialVideo",
                " OfficialMusicVideo", "OfficialAudio", "(Official Music Video)", "(Official Audio)", "(Official Video)"]
pat = r'\b(?:{})\b'.format('|'.join(remove_words))
df_song['Song'] = df_song['Song'].str.replace(pat, '')
df_song['Song'] = df_song['Song'].str.replace("[", "")
df_song['Song'] = df_song['Song'].str.replace("]", "")
single_clean = df_song.to_csv("clean_song.csv")

/var/folders/pm/v25r7xvn0lvgn9vffsrn5nxw0000gn/T/ipykernel_46049/2232784837.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  df_song['Song'] = df_song['Song'].str.replace(pat, '')
/var/folders/pm/v25r7xvn0lvgn9vffsrn5nxw0000gn/T/ipykernel_46049/2232784837.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_song['Song'] = df_song['Song'].str.replace(pat, '')
/var/folders/pm/v25r7xvn0lvgn9vffsrn5nxw0000gn/T/ipykernel_46049/2232784837.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_song['Song'] = df_song['Song'].str.replace("[", "")

Now i'm going to clean the album. I got 2 options, one is easy and the other one is harder.
- the first option is only to collect the name of the album in the data
- the second one is to collect the name of every single from the data